library/std/src/sync/rwlock/tests.rs

```diff
@@ -501,3 +501,143 @@ fn panic_while_mapping_write_unlocked_poison() {
 
     drop(lock);
 }
+
+#[test]
+fn test_downgrade_basic() {
+    let r = RwLock::new(());
+
+    let write_guard = r.write().unwrap();
+    let _read_guard = RwLockWriteGuard::downgrade(write_guard);
+}
+
+#[test]
+fn test_downgrade_frob() {
+    const N: u32 = 10;
+    const M: usize = if cfg!(miri) { 100 } else { 1000 };
+
+    let r = Arc::new(RwLock::new(()));
+
+    let (tx, rx) = channel::<()>();
+    for _ in 0..N {
+        let tx = tx.clone();
+        let r = r.clone();
+        thread::spawn(move || {
+            let mut rng = crate::test_helpers::test_rng();
+            for _ in 0..M {
+                if rng.gen_bool(1.0 / (N as f64)) {
```

Comment:
Could this just use something like `N % 2` so the test is less non-deterministic?

This test could use a comment, I'm not sure what the purpose of the channel is unless it's a workaround for collecting handles to a vec and joining.

Reply:
Honestly I think I should remove this test altogether, it is simply a copy of the test above (`frob`) but with downgrading included

Reply:
Ah yes it seems like a120fb7 removes the frob test altogether so I'll get rid of this.